### PR TITLE
[WIP] Fix many protocol handling issues and extend protocol support

### PR DESCRIPTION
**Pull request by gwynne**
This is a work in progress draft for fixing up many long-standing problems with the wire protocol handling in MySQLNIO, with an eye towards improving performance (most of what's been done so far at the time of this writing lays groundwork for said improvements).

This PR causes some public API breakages, but limits said breakages to APIs that should never have been public (and for the most part are not accessible publicly regardless). Some minor fixups in MySQLKit will be included to insulate clients from the changes where it is deemed necessary.

### CODE REVIEW

File: Sources/MySQLNIO/Commands/MySQLQueryCommand.swift [hunk 1]
```diff
@@ -161,6 +162,10 @@ private final class MySQLQueryCommand: MySQLCommand {
             self.state = .executeColumns(remaining: numericCast(count))
             return .noResponse
         case .executeColumns(var remaining):
+            if self.ok!.numColumns == 0 && remaining > 0 {
```
Comment (0xTim) [inline Sources/MySQLNIO/Commands/MySQLQueryCommand.swift:165]: From my reading of the code, where does `ok` get set? The force unwrap seems odd with no checks above

**Review by 0xTim — COMMENTED** [review]
LGTM, a couple of queries. 

At a more general comment - there are a lot of breaking changes here. It tidies up the API nicely. I'm not adverse to releasing this as a major version if we decide the threshold is too high for the breaking changes

File: Sources/MySQLNIO/Commands/MySQLQueryCommand.swift (hunk 1 above)
Comment (gwynne) [reply Sources/MySQLNIO/Commands/MySQLQueryCommand.swift:165]: This is set earlier in the state machine, if it hasn't been set by the time this state is reached, the state machine itself is incorrect (i.e. it's legitimately programmer error; that code path can't be induced by a misbehaving server). The whole state machine is a mess, though, and cleaning it up further is on the remaining todo list for this PR.

**Comment by gwynne** [comment]
> LGTM, a couple of queries.
> 
> At a more general comment - there are a lot of breaking changes here. It tidies up the API nicely. I'm not adverse to releasing this as a major version if we decide the threshold is too high for the breaking changes

I've tried to limit the API-breaking changes to things that will only break users who were using the package incorrectly in the first place - most of the packet types already couldn't be usefully instantiated from outside the package despite being `public`, for example. For the case of `ERR_Packet`, I have a local branch on `mysql-kit` which will `typealias` the new name for anyone who was relying on the original via `MySQLError`.

**Comment by gwynne** [comment]
For those following this PR, please be assured it is not being abandoned; the work has been moved to a new branch that will be available shortly.

**Pull request closed without merging**